### PR TITLE
Support strict_exception_groups=True

### LINKED ISSUE #132: Hide MultiErrors
Several of the APIs in this library create a "hidden" nursery that runs the connection's background task. This has the surprising side effect that something that looks like single-task code can actually raise a `MultiError`. Nathaniel has a good thread describing this. I'll just steal a snippet to illustrate the problem here:

```python3
async with open_websocket("https://...") as ws:
    raise ValueError
```

> but if `open_websocket` has a nursery hidden inside it, then you'll get a `MultiError([ValueError])` instead of a `ValueError`, and that's going to surprise people, because it's mostly an implementation detail that `open_websocket` creates a nursery.

The proposed solution is to design the library to avoid raising exceptions on the background task. This would allow us to catch `MultiError` in the hidden nursery, unwrap the nested exception, and raise it to the caller. If the background task does raise an error, then we wrap* the entire `MultiError` inside of a new class `TrioWebsocketInternalError`. This wrapper class is a signal to the caller that the bug is in this library, not in their code.

*Nathaniel's post suggests that the internal error either is-a or has-a `MultiError`. We can also look at `TrioInternalError` for inspiration.

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes #132 and #187

* Changes `open_websocket` to only raise a single exception, even when running under `strict_exception_groups=True`
  * [ ] Should maybe introduce special handling for `KeyboardInterrupt`s
  * If multiple non-Cancelled-exceptions are encountered, then it will raise `TrioWebSocketInternalError` with the exceptiongroup as its `__cause__`. This should only be possible if the background task and the user context both raise exceptions. This would previously raise a `MultiError` with both Exceptions.
  * other alternatives could include throwing out the exception from the background task, raising an ExceptionGroup with both errors, or trying to do something fancy with `__cause__` or `__context__`. 
* `WebSocketServer.run` and `WebSocketServer._handle_connection` are the other two places that opens a nursery. I've opted not to change these, since I don't think user code should expect any special exceptions from it, and it seems less obscure that it might contain an internal nursery.
  * [ ] Update docstrings to mention existence of internal nursery.

### CODE REVIEW

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:1]: Hi John-- I may be slow in reviewing.

Please make sure the PR description summarizes the changes.

Would need test coverage for any changes, but ok to hold off until I've reviewed the basics (as I can't promise we'll take the direction of this PR).

> I'm itching to add Black and type checking to the CI, but that should be in a separate PR.

Please see #177.   On type checking, I'm ok running a loose mypy pass, but reluctant to convert the codebase to strict type annotations, since fighting the type system is often a burden on maintenance (especially with a project on life support).

Comment (jakkdl) [reply trio_websocket/_impl.py:1]: Ah thank you, of course I'd complained about lack of Black before :sweat_smile: 

Agree on loose mypy just to check if the annotations that exist are correct. And I don't have any plans on adding annotations everywhere, just doing it for my own sake when developing.

Will try and make PR summary verbose and descriptive :+1:

**Comment by jakkdl** [comment]
@belm0 do you have strong opinions on keeping python3.7 support? It's getting problematic to support with lotsa libraries not supporting it. If pinning test requirements to versions that support 3.7 we're not testing anything remotely recent, so would need to generate an additional requirements.txt file, etc etc.

Testing trio versions < 0.22 is also getting messy. If we explicitly want to check for `strict_exception_group` compatibility the parameter was only introduced in 0.22
I would also love to use `trio.testing.RaisesGroup`, but at that point the test deps hit trio>=0.24

**Comment by belm0** [comment]
> @belm0 do you have strong opinions on keeping python3.7 support? It's getting problematic to support with lotsa libraries not supporting it.

I'm ok dropping 3.7 since it's past EOL.   (For what it's worth, mainstream packages now think it's ok to drop versions before EOL-- numpy and others dropping 3.8 already.)

> Testing trio versions < 0.22 is also getting messy. If we explicitly want to check for `strict_exception_group` compatibility the parameter was only introduced in 0.22 I would also love to use `trio.testing.RaisesGroup`, but at that point the test deps hit trio>=0.24

My org uses trio < 0.22 and would like to continue using the latest trio-websocket (otherwise I'm no longer using the package I maintain).  My view is that trio's handling of the strict exception group transition is problematic (especially making strict_exception_group a runtime option that can have effect globally on transitive dependencies), so we're just avoiding it.

**Comment by jakkdl** [comment]
did some messing around:
* Doing `shield = True` in `close_connection` is probably a bad idea when the `disconnect_timeout` defaults to a full minute. It might be a more proper thing to do, but that'd be a change of behavior that this PR does not care to mess with.
* I looked at KeyboardInterrupt, and if a user were to cause one while `_reader_task` happens to be the current task, I think it will cause a `TrioWebsocketInternalError`. Would need to do signal handling in order to test it, or monkeypatch `_reader_task` or one of the handlers it uses to trigger this.

File: tests/test_connection.py
Comment (belm0) [inline tests/test_connection.py:141]: Python < 3.9 doesn't support direct use of `list`, etc. for type annotations

File: Makefile
Comment (belm0) [inline Makefile:15]: I usually put type checking on a separate rule `type_check`

File: tests/test_connection.py
Comment (belm0) [inline tests/test_connection.py:142]: Is this a mistake?  I don't know why the lambda return expression includes a type annotation (`[str]`).

File: tests/test_connection.py
Comment (belm0) [inline tests/test_connection.py:446]: for cases like this, I would rather see an explicit `assert exc` in the code over a pylint annotation

File: tests/test_connection.py
Comment (belm0) [inline tests/test_connection.py:462]: please move import to the module scope

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:51]: needs docs

> If multiple non-Cancelled-exceptions are encountered, then it will raise TrioWebSocketInternalError with the exceptiongroup as its __cause__. This should only be possible if the background task and the user context both raise exceptions. This would previously raise a MultiError with both Exceptions.

> other alternatives could include throwing out the exception from the background task, raising an ExceptionGroup with both errors, or trying to do something fancy with __cause__ or __context__.

It seems odd to raise an internal error where a user-context exception is implicated.  Perhaps we can discard the internal error with a warning message, and propagate the user exception.

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:753]: using pipe for union is python 3.10

doesn't mypy complain about these if we run the type check under 3.8?

File: setup.py
Comment (belm0) [inline setup.py:31]: I'd like to move dropping python 3.7 and adding mypy in separate PR(s) and merge soon, then rebase to that.  I may give it a try.

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:1]: > WebSocketServer.run and WebSocketServer._handle_connection are the other two places that opens a nursery. I've opted not to change these, since I don't think user code should expect any special exceptions from it, and it seems less obscure that it might contain an internal nursery.

In principle I'd like to state that the API doesn't raise exception groups, and that would include the server API.  For the user's handler, perhaps add `UserRequestHandlerException`?

File: tests/test_connection.py
Comment (jakkdl) [reply tests/test_connection.py:462]: The import is only possible on trio>0.25, i.e. if `not _trio_default_loose()`. But I suppose I can call that on the module scope, or do a `try: ... except ImportError`

File: trio_websocket/_impl.py
Comment (jakkdl) [reply trio_websocket/_impl.py:753]: It works because I added `from __future__ import annotations`

File: tests/test_connection.py
Comment (jakkdl) [reply tests/test_connection.py:141]: It works because I added `from __future__ import annotations`, otherwise the CI test would fail.

File: setup.py
Comment (jakkdl) [reply setup.py:31]: I'll likely be busy the rest of the week, but can do it afterwards if you haven't gotten around to it.

File: trio_websocket/_impl.py
Comment (jakkdl) [reply trio_websocket/_impl.py:1]: I think that for users that are embracing `strict_exception_groups=True` and are used to handling exceptiongroups everywhere we shouldn't purposely obscure the groups and make errors harder for them to handle (and Trio's stance is that this is the default and should be embraced). I haven't delved too too deeply into them to see what it would imply to try and avoid exceptiongroups, but I think it might necessitate a change in API where we might end up changing what exceptions are being raised - even for users sticking with `strict_exception_groups=False`.

File: tests/test_connection.py
Comment (jakkdl) [reply tests/test_connection.py:446]: it's `BaseExceptionGroup` that pylint fails to realize will always be available. But can switch to an `assert`

File: tests/test_connection.py
Comment (jakkdl) [reply tests/test_connection.py:142]: the annotation is for `trio.open_memory_channel`, specifying the type of the objects being sent over the channels. It's not normally possible to specify the type on functions, so Trio says it's a callable class during type checking to allow for that usage: https://github.com/python-trio/trio/blob/4dcdc6dbe899c9ee295fa9c8e6a077100ae19f16/src/trio/_channel.py#L91

Without the annotation, `queued_streams` would need the same long & verbose type annotation as `accepted_streams`.

File: tests/test_connection.py
Comment (jakkdl) [reply tests/test_connection.py:446]: ...except mypy correctly picks up that it will always be truthy, and raises a warning if doing `assert BaseExceptionGroup` >.>

**Comment by jakkdl** [comment]
I rewrote the history with some squashing rebases, so even without merging #189 one can now selectively review this PR by only viewing 7eb779b in the "Files Changed" tab.

**Comment by jakkdl** [comment]
>     * I looked at KeyboardInterrupt, and if a user were to cause one while `_reader_task` happens to be the current task, I think it will cause a `TrioWebsocketInternalError`. Would need to do signal handling in order to test it, or monkeypatch `_reader_task` or one of the handlers it uses to trigger this.

Okay I think I'm finally starting to understand `KeyboardInterrupt`. If either `_reader_task` or the user code raises `KeyboardInterrupt`, causing the other to get `Cancelled`, then we simply throw away the `Cancelled` and raise the original `KeyboardInterrupt`. But the more complex problem is if the user code raises some exception, and before we finish unwinding `_reader_task` and/or during `close_connection` we get a `KeyboardInterrupt`. At this point we definitely shouldn't suppress KI and just raise a warning, in favor of raising the users code. This is a prime example of when it would be so much easier to always raise a group, but I think what we ultimately have to do is `raise ki_exception from user_exception`.

**Comment by jakkdl** [comment]
Big rewrite, fairly happy with the state of things - except:
1. the multiple added tests are very similar and could probably have common code moved into helpers/be parametrized/etc.
2. I had trouble engineering `DisconnectionTimeout`s, so there's no tests for exception chaining into that. I could of course monkeypatch that to happen though.


debugging old_deps is nooooo fun when the only exception being surfaced to me is errors inside TracebackException. :upside_down_face:

File: requirements-extras.in
Comment (belm0) [inline requirements-extras.in:4]: mypy is listed twice

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:182]: It's repeated.   Consider renaming `_TRIO_MULTI_ERROR` -> `_IS_TRIO_MULTI_ERROR`, and have `_TRIO_MULTI_ERROR` be the multi-error class itself (MultiError or BaseExceptionGroup)

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:185]: I forgot if we discussed already, but I think `Optional[Foo]` is more friendly / readable than `Foo | None`.

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:242]: perhaps "The trio-websocket API is not expected to raise multiple exceptions"

File: trio_websocket/_impl.py
Comment (jakkdl) [reply trio_websocket/_impl.py:185]: I think `|None` is the standard way of doing it nowadays, and pyupgrade/ruff will warn on doing `Optional[...]`: https://docs.astral.sh/ruff/rules/non-pep604-annotation/

I also think it's nice for newbies not to have another keyword to know, and highly personally I prefer not having to do another import for it.

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:153]: Multiple Cancelled was an existing possible case, so do we need to try hard to make it a single exception?   (Cancelled is typically caught by trio itself, and otherwise the user had to deal with MultiError Cancelled anyway.)

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:164]: please name these as `_open/_close_connection()` to clarify that they're internal

File: trio_websocket/_impl.py
Comment (belm0) [inline trio_websocket/_impl.py:185]: consider `result2` -> `close_result`

File: trio_websocket/_impl.py
Comment (jakkdl) [reply trio_websocket/_impl.py:153]: The problem is that it's a very easy gotcha to write
```python
try:
  ...
except Cancelled:
  ...
```
which might work 99% of the time, but then at some random point you get multiple instead of a single (or the other way around) `Cancelled`, and your exception handler is now broken. I agree that `Cancelled` is perhaps a special case due to trio handling it, but if it does surface to the user I'm not sure there's much to be gained from having them see multiple of them.

File: tests/test_connection.py
Comment (jakkdl) [reply tests/test_connection.py:527]: looks like trio docs say "non-strict", so we can follow that: https://trio.readthedocs.io/en/stable/reference-core.html#historical-note-non-strict-exceptiongroups

File: trio_websocket/_impl.py
Comment (belm0) [reply trio_websocket/_impl.py:153]: MultiError with Cancelled can happen fairly easily, and we quickly learned never to write `except Cancelled`.

By this trio-websocket change, we're masking the user from the effects of a trio anti-pattern, which is concerning.   Like, the user may get away with `except Cancelled` when the trio-websocket API is involved, but then be doubly puzzled when it doesn't work with another API.

Comment (jakkdl) [reply trio_websocket/_impl.py:153]: Hrm, I think this would be a very bad design choice. With this PR we're trying to make the internal nursery an implementation detail, as per https://trio.readthedocs.io/en/stable/reference-core.html#designing-for-multiple-errors
Before `strict_exception_groups=True` was the default it was expected that a trio developer would be able to handle the occasional `MultiError`, but since `strict=True` trio is trying very hard to make it fully consistent - either you always get an `ExceptionGroup`, or you never get one. And all other library APIs are encouraged to follow in those footsteps.

In the future the users shouldn't need to worry about this anti-pattern at all, so I do not think we should enshrine it in this API.

File: trio_websocket/_impl.py
Comment (jakkdl) [reply trio_websocket/_impl.py:185]: much better, thanks

**Review by belm0 — APPROVED** [review]

**Comment by fossdd** [comment]
hmm, i still got with this patch, failures running the `streamlink` testsuite...:

```
=============================================== short test summary info ===============================================
FAILED tests/test_api_validate.py::TestParseQsdValidator::test_failure - assert "ValidationEr...t 'int' (123)" == "ValidationEr...decode' (123)"
FAILED tests/webbrowser/cdp/test_client.py::TestEvaluate::test_exception - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_client.py::TestEvaluate::test_error - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_client.py::TestNavigate::test_detach - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_client.py::TestNavigate::test_error - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestCreateConnection::test_failure - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestReaderError::test_invalid_json - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestReaderError::test_unknown_session_id - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestSend::test_timeout[Default timeout, response not in time] - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestSend::test_timeout[Custom timeout, response not in time] - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestSend::test_bad_command - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestSend::test_result_exception - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestHandleCmdResponse::test_response_error - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/cdp/test_connection.py::TestHandleCmdResponse::test_response_no_result - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/webbrowser/test_webbrowser.py::TestLaunch::test_terminate_on_nursery_baseexception - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
=============================== 15 failed, 6375 passed, 28 skipped, 1 warning in 14.93s ===============================
```

**Comment by bastimeyer** [comment]
Apologies for the off-topic comment, but since I'm the maintainer of Streamlink, I'll have to answer the previous comment.

> failures running the `streamlink` testsuite

**`trio-websocket` is not actually involved in any of the tests.**

Streamlink's webbrowser tests fully mock/stub/monkeypatch `trio_websocket.connect_websocket_url` / `WebSocketConnection`, because the tests are only interested in message handling, not **how** the messages are sent or received.

- https://github.com/streamlink/streamlink/blob/6.10.0/tests/webbrowser/cdp/test_connection.py#L173
- https://github.com/streamlink/streamlink/blob/6.10.0/tests/webbrowser/cdp/conftest.py#L8-L18
- https://github.com/streamlink/streamlink/blob/6.10.0/tests/webbrowser/cdp/__init__.py#L10-L33

Your test failures are caused by something else. Considering that other tests are failing as well, you should probably have a look at how you package `trio`. The Streamlink tests have been running just fine since the breaking changes of trio 0.25 were resolved. See [GH actions](https://github.com/streamlink/streamlink/actions/runs/10968036551) running everything in a clean venv using the latest dependency versions.

- https://github.com/streamlink/streamlink/commit/1a7295b110777473575f41b868492fd56628b7df
- https://github.com/streamlink/streamlink/commit/78d0b58d757ada9049667dee3a7369d4e6b6dc9d
- https://github.com/streamlink/streamlink/commit/d5aa26f6f5d8290fef55dd0e11670d9e9096973d

> `tests/webbrowser/test_webbrowser.py::TestLaunch::test_terminate_on_nursery_baseexception`

This test is part of an outdated version (changed in `d5aa26f` / `6.7.3`).

**Pull request merged**
